crates/wysiwyg/src/composer_model/code_block.rs

```diff
@@ -435,149 +247,114 @@ where
-        let path_len = handle.depth();
-        if path_len <= 1 {
-            DomHandle::root()
-        } else {
-            for i in (0..handle.depth()).rev() {
-                let ancestor_handle = handle.sub_handle_up_to(i);
-                let ancestor = self.state.dom.lookup_node(&ancestor_handle);
-                match ancestor.kind() {
-                    Generic | List | ListItem | CodeBlock => {
-                        return ancestor_handle
-                    }
-                    _ => continue,
-                }
-            }
-            panic!("Should never reach this point, one of the parents surely can be split.");
-        }
-    }
-
-    pub(crate) fn format_node_for_code_block(node: &DomNode<S>) -> DomNode<S> {
+    /// Converts any nodes to be added to a code block to the right format, recursively.
+    /// Line breaks get turned into `\n` chars.
+    /// Text nodes are just cloned.
+    /// Container nodes will be added to the code block and their contents will also be formatted:
+    /// Block nodes and list items will add extra line breaks (`\n` chars).
+    pub(crate) fn format_node_for_code_block(
+        &mut self,
+        node: &DomNode<S>,
+        range: &Range,
+        first_leaf: &DomLocation,
+        last_leaf: &DomLocation,
+    ) -> Vec<DomNode<S>> {
+        // TODO: try to diff node positions and offsets in a more straightforward way
+        // These checks are used to verify if the changes in the Dom should also move the selection
+        let is_in_selection = range.contains(&node.handle());
+        let is_before =
+            node.handle() <= first_leaf.node_handle && !is_in_selection;
+        let is_after =
+            node.handle() > last_leaf.node_handle && !is_in_selection;
         match node {
             DomNode::LineBreak(_) => {
+                // Just turn them into line break characters
                 let mut text_node = DomNode::new_text("\n".into());
-                text_node.set_handle(node.handle());
-                text_node
+                text_node.set_handle(node.handle().clone());
+                vec![text_node]
             }
-            DomNode::Text(_) => node.clone(),
+            DomNode::Text(_) => vec![node.clone()],
             DomNode::Container(container) => {
                 let mut children = Vec::new();
+                // We process each child node
                 for c in container.children() {
-                    children.push(Self::format_node_for_code_block(c));
+                    children.extend(self.format_node_for_code_block(
+                        c, range, first_leaf, last_leaf,
+                    ));
                 }
-                DomNode::Container(container.clone_with_new_children(children))
+                // Then post-process the containers to add line breaks and move selection if needed
+                if container.is_block_node() || container.is_list_item() {
+                    if container.is_list() {
```

Comment:
the content from 287 to  338 could be a function of it's since it's very nested (and maybe some of it's own nested pieces may be other functions?)

Reply:
Done in 3b957ad1a90e53360a56816dde413d8954ee7a98